### ISSUE TITLE
Cannot get HubCallerContext.User to get Set

### ISSUE THREAD

**Issue by chris-rogala** [dotnet/aspnetcore#36421]
### Describe the bug
I have an web application that I am trying to setup websockets for. The user has a customerId as a claim and the Hub is designed to send messages for all connections with the customerId. The web application authenticates through the [Auth0](https://auth0.com/) identity provider. When the authentication occurs, the AuthenticationType is set to Federation. The HttpContext in the controller recognizes the Authenticated user. I was also able to pull the user in the view and the component, so I know the HttpContext does have the User properly set. However, if I add the Authorize attribute to the Hub class, then I receive a 401. If I remove the attribute, the HubCallerContext.User is not the authenticated user. Since this is a web application that is interacting with a Hub inside the application, the Hub should recognize the User.

### To Reproduce
As this requires an identity provider, and the identity provider I use is a paid identity provider, I don't have a repo. That said, this should be a fairly straightforward issue. Here is a link to a stackoverflow question that details the steps to reproduce.

### Exceptions (if any)
I receive a 401 if the Authorize attribute is applied to the hub. 

### Further technical details
- ASP.NET Core: 3.1
- Include the output of `dotnet --info`: N/A
- The IDE (VS / VS Code/ VS4Mac): VS2019

**Comment by adityamandaleeka** [comment]
Triage: have you reached out to Auth0 about this issue? This feels like a nuance of their auth provider with which we don't have much familiarity.

BTW, have you validated that the access token that you're passing to the component is a JWT token? You can check it using https://jwt.ms/

**Comment by chris-rogala** [comment]
@adityamandaleeka Here is a very basic repo. To create it, I used the Web Application (MVC) template with the Authentication using Azure AD. I have directions in the readme.md file that walks through the scenarios. I hope this helps: https://github.com/chris-rogala/Microsoft.SignalR.Federated

**Comment by BrennanConroy** [comment]
You aren't passing an access token through the HubConnection in your sample, also you need to be aware that the token might be in the query string instead of the header, check out the doc about jwt tokens https://docs.microsoft.com/aspnet/core/signalr/authn-and-authz?view=aspnetcore-5.0#bearer-token-authentication

**Comment by chris-rogala** [comment]
@BrennanConroy That is correct. Based on this [documentation](https://docs.microsoft.com/en-us/aspnet/core/signalr/authn-and-authz?view=aspnetcore-3.1), the cookie should be passed to the hub:

> In a browser-based app, cookie authentication allows your existing user credentials to automatically flow to SignalR connections. When using the browser client, no additional configuration is needed. If the user is logged in to your app, the SignalR connection automatically inherits this authentication.

I added the access token to see if it would be passed to the component. I realize that the access token was issued with the audience of the web API and should not work. I would have to move the Hub to the Web API, but I would prefer for the Hub to stay within this web application. The current flow is for the Web API to send messages out to web applications that are subscribed to the events and then the web app sends the information back to its clients with SignalR.

I would think that the context would be set since the client is interacting with the web app directly, and has a session. Let me know if you have any additional questoins.

**Comment by BrennanConroy** [comment]
You aren't using a browser-based app though. Because you are using server-side blazor your components code runs **on the server** which means the HubConnection you create isn't in the browser thus doesn't automatically flow the cookie.

I found a nice blog post that shows steps to use cookie auth with server-side blazor https://blazorhelpwebsite.com/ViewBlogPost/36. It uses HttpClient, but you should be able to do a similar thing with SignalR.

**Comment by chris-rogala** [comment]
@BrennanConroy Are you saying that the SignalR package is not compatible with Blazor?

**Comment by BrennanConroy** [comment]
I did not say that, I said that your assumption that cookies flow automatically isn't accurate because it's a server-side app.

**Comment by chris-rogala** [comment]
@BrennanConroy I am a bit confused on the recommended pattern for SignalR with respect to Blazor. In my example, I have a simple server-side application, which is still extremely common in the industry. In all previous iterations of .NET applications, the server-side application managed the identity through cookies. In this application, the identity is also managed through a cookie. See the screenshot below:

![image](https://user-images.githubusercontent.com/44238629/133347451-be20334e-4991-43f3-b6c4-dc61c74e8d16.png)

From my understanding, Blazor is supposed to allow a developer to write client-side code in C# instead of javascript. In my scenario, the federated service authenticates the user and then manages the users session with a cookie which you can see above. You can also verify this in the browser:

![image](https://user-images.githubusercontent.com/44238629/133348360-0530ad97-9f6a-480b-b7c6-d6f40517d4a3.png)

In the example you provided, this follows a cookie authentication pattern that is not based on OpenID. While these are similar, they are completely different and the cookies are not the same since this cookie starts with .AspNetCore.AzureADCookie and it is separated into two chunks. You can verify this by running the application locally.

Additionally, while trying to get the application to run in the unsecure Hub, I ran into the same error that was mentioned here [https://github.com/dotnet/aspnetcore/issues/34419](https://github.com/dotnet/aspnetcore/issues/34419). I know you requested a server log, but running this repo locally should provide the server log and allow you to reproduce the error. I know this was closed until more people run into this issue, but the reality is, not all developers report the issue and they just simply move out to a different technology or adjust their pattern to make it work.

Please note, at the end of that issue, a user posted a valid question which is inline with my questions:

> @mkArtakMSFT, @BrennanConroy what is the typical pattern seen here? Would the SignalR Hub run in a separate app, secured via JWTs or otherwise and the Blazor client acquire an access token to the Hub? I'm trying to balance my understanding of the pieces because it gets more complex and "interesting" when things require Authentication -- and even more interesting when your app(s) are multi-tenant and require consent, etc. If I wasn't concerned with who a user is and what they're trying to do, I'd just throw apps and services around like confetti but in these scenarios, confetti is reserved for when you ship 😉
> 
> Appreciate the thoughts

While I can move the Hub to a Web API and then pass the access token to that endpoint, I do find this very disheartening that this package and/or Blazor does not appear to integrate easily with Azure AD. I am only going to have more and more of my clients moving to Azure AD or a similar Federated service, and I am sure I am not alone.

Thanks in advance for your feedback.

**Comment by chris-rogala** [comment]
@BrennanConroy I pushed out some changes to the components so you can verify the authentication state in the component before the connection is attempted. Line 20:

```
var state = await AuthenticationStateProvider.GetAuthenticationStateAsync()
                .ConfigureAwait(false);
```

In the component, it results in the state having the user. 

![image](https://user-images.githubusercontent.com/44238629/133350164-790debc9-da21-444b-8d4a-a02778585dde.png)

**Comment by BrennanConroy** [comment]
> Blazor is supposed to allow a developer to write client-side code in C# instead of javascript

Sort of. There are two separate models here.

There is blazor server-side, which is a bit magic and looks like it's running code in the browser but is actually running it all on the server-side. The problem with that is you don't get things the browser does for you like cookies flowing automatically because you're actually running the code on the server and not the browser. You keep pointing out that controllers "just work", but this is because you're using them outside of the blazor context, so the browser is making an http call to the controller which means you get the nice browser behavior of cookies etc. If you instead made an http call from your component to the controller, you would see the same issue you're seeing with the HubConnection where it isn't authenticated. This is because your "client" code is now running on the server instead of in the browser.

The other mode, Blazor WebAssembly (WASM), will actually run your component code in the browser, so if you made an http (or SignalR) call in your component then cookies should start working automatically because the http call is now being made from the browser instead of from the server.

There are pros and cons to using either model and you are hitting one of the cons with server-side blazor.

You can read about the hosting models at https://docs.microsoft.com/en-us/aspnet/core/blazor/hosting-models?view=aspnetcore-5.0

There is also a doc about passing tokens to a component in server-side blazor for an http call https://docs.microsoft.com/aspnet/core/blazor/security/server/additional-scenarios?view=aspnetcore-5.0

**Comment by chris-rogala** [comment]
@BrennanConroy Thank you for the links and the explanation. That being said, why does it appear that SignalR is allowed in the server-side scenario?

> There is blazor server-side, which is a bit magic and looks like it's running code in the browser but is actually running it all on the server-side. The problem with that is you don't get things the browser does for you like cookies flowing automatically because you're actually running the code on the server and not the browser.

Since this code is running server-side, what is the point for SignalR to even work in this scenario? Does this create a web-socket from one web application to another web application? If so, wouldn't that cause a socket exhaustion within either application? Is this by design and even applicable? And if this does work, how is this beneficial? Technically the connection would be from server to server with no user connection and therefor the client would never receive a call from the web application.

I'll try to modify the example repo based on your links you mentioned above. While I do that, would either you or @mkArtakMSFT be able to provide more details around the question I referenced?

> @mkArtakMSFT, @BrennanConroy what is the typical pattern seen here? Would the SignalR Hub run in a separate app, secured via JWTs or otherwise and the Blazor client acquire an access token to the Hub? I'm trying to balance my understanding of the pieces because it gets more complex and "interesting" when things require Authentication -- and even more interesting when your app(s) are multi-tenant and require consent, etc. If I wasn't concerned with who a user is and what they're trying to do, I'd just throw apps and services around like confetti but in these scenarios, confetti is reserved for when you ship 😉
> 
> Appreciate the thoughts

Thanks!

**Comment by chris-rogala** [comment]
@BrennanConroy I did verify that the connection being made is from the server to itself, which appears to be pointless in the scenario I am trying to work with. I'm sure there is a reason for this, but it does not jump out at me.

As for Blazor, at first glance, Blazor reminds me of Silverlight, and that didn't go well for Microsoft. I looked into the WebAssembly option, but I don't like the idea of having to download a framework to the browser. Again, this seems to be very similar to Silverlight and turned bad years ago.

For now, I am going to go back to javsscript to create the connection. It would be really cool if Blazor would handle this and prerender the HTML so it is sent to the client, but I guess that would have to be a future feature. Best of luck and thank you for the feedback.
